glos_qartod/cli.py

```diff
@@ -30,36 +32,98 @@ def main():
+    run_qc_list(config, args.netcdf_files)
 
-def run_qc(config, ncfile):
+def run_qc_list(config, nc_file_list):
+    for nc_path in nc_file_list:
+        with Dataset(nc_path, 'r') as nc:
+            run_qc(config, nc)
+
+def open_and_qc(config, nc_path):
+    with Dataset(nc_path, 'r') as nc:
+        run_qc(config, nc)
+
+
+def create_or_open_qc_file(qc_file_name, parent_dimensions):
+    """
+    Given a name of a QC file, attempt to open it.  If it exists, check
+    that it has the same dimensions as "dimensions", primarily taken from
+    the raw parent NetCDF dataset and then open it.  If it does not, create a
+    new NetCDF file with the specified dimensions.  Returns an opened netCDF
+    Dataset object.
+
+    :param qc_file_name str: The path of the file to open or create
+    :param dimensions list: A list of tuples with the dimensions
+    """
+    if os.path.exists(qc_file_name):
+        try:
+            ncfile = Dataset(qc_file_name, 'a')
+            if ncfile.dimensions != parent_dimensions:
+                raise ValueError("File {} had dimensional mismatch with original data dimensions, recreating QC file".format(qc_file_name))
+            else:
+                return ncfile
+            # if an exception is returned, log it and fall through to write the
+            # QC file from scratch
+        except Exception as e:
+            ncfile.close()
+            get_logger().warning(str(e))
+
+    # if we reached here, either the qc file didn't exist or ran into an error
+    # while trying to be opened, so attempt to create the file from scratch
+    ncfile = Dataset(qc_file_name, 'w')
+
+    # create dimensions
+    for d in six.itervalues(parent_dimensions):
+        ncfile.createDimension(d.name, d.size)
+
+    return ncfile
+
+
+
+
+def run_qc(config, ncfile, qc_suffix=None):
     '''
     Runs QC on a netCDF file
     '''
-    qc = DatasetQC(ncfile, config)
-    for varname in qc.find_geophysical_variables():
-        get_logger().info("Applying QC to %s", varname)
-        ncvar = ncfile.variables[varname]
-        for qcvarname in qc.create_qc_variables(ncvar):
-            qcvar = ncfile.variables[qcvarname]
-            get_logger().info(qcvarname)
-            qc.apply_qc(qcvar)
-        get_logger().info("Primary QC")
-        qc.apply_primary_qc(ncvar)
+    qc_end = 'ncq' if qc_suffix is None else qc_suffix
+    fname_base, ext = os.path.splitext(ncfile.filepath())
+    qc_filename = fname_base + '.{}'.format(qc_end)
+    # Look for existing qc_file or return new one
+    qc_file = create_or_open_qc_file(qc_filename, ncfile.dimensions)
+    # load NcML aggregation if it exists
+    ncml_filename = fname_base + '.ncml'
+    qc = DatasetQC(ncfile, qc_file, ncml_filename, config)
+    # zero length times will throw an IndexError in the netCDF interface,
+    # and won't result in any QC being applied anyways, so skip them if present
+    if len(qc.ncfile.variables['time']) > 0:
```

Comment:
Ok, in this case we'll stick with len. The definition of the len operator is generally ambiguous because the netCDF folks could implement it or change it at their whim. For all I know, their definition of `len` could be that the variable name contains characters...